### PR TITLE
[SPARK-44162][CORE] Support G1GC in spark metrics

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
As a part of support JDK21, add support `G1 Concurrent GC` GarbageCollectorMXBean in Spark metrics.
Refer https://github.com/openjdk/jdk/pull/11341 and https://bugs.openjdk.org/browse/JDK-8297247 , the `G1 Concurrent GC` not a part of YoungGC or FullGC. So we follow the JDK definition, bring two new metrics: `ConcurrentGCCount` and `ConcurrentGCTime`.

### Why are the changes needed?
add new builtin garbage collectors for metrics.

### Does this PR introduce _any_ user-facing change?
Yes, will receive new metrics.

### How was this patch tested?
Test in local, I will add test if necessary.

### CODE REVIEW

**Comment by Hisoka-X** [comment]
@dongjoon-hyun @LuciferYang Hi, PTAL. By the way, which test files should I update? Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @Hisoka-X . Could you re-trigger the failed pipelines?
```
[info] *** 43 TESTS FAILED ***
[error] Failed: Total 3660, Failed 43, Errors 0, Passed 3617, Ignored 10, Canceled 4
[error] Failed tests:
[error] 	org.apache.spark.deploy.history.RocksDBBackendHistoryServerSuite
[error] 	org.apache.spark.util.JsonProtocolSuite
[error] 	org.apache.spark.deploy.history.LevelDBBackendHistoryServerSuite
```

**Comment by Hisoka-X** [comment]
@dongjoon-hyun Hi, I follow the guide to regenerated the test file in https://github.com/apache/spark/tree/master/core/src/test/resources/HistoryServerExpectations by `HistoryServerSuite::main`, many json style changed. Seem like other developer like change test file manually, not regenerate it. Should I revert these change or keep it?

File: core/src/test/resources/HistoryServerExpectations/application_list_json_expectation.json [hunk 1]
```diff
@@ -54,8 +54,8 @@
     "sparkUser" : "tgraves",
     "completed" : true,
     "appSparkVersion" : "3.0.0-SNAPSHOT",
-    "endTimeEpoch" : 1578764802615,
     "startTimeEpoch" : 1578764662851,
+    "endTimeEpoch" : 1578764802615,
```
Comment (srowen) [inline core/src/test/resources/HistoryServerExpectations/application_list_json_expectation.json:58]: I don't think we have to do it here, but it'd be nice to sort the output keys. These diffs are almost all just lines moving around

**Comment by dongjoon-hyun** [comment]
Hi, @Hisoka-X . Could you address the previous review comment first?

**Comment by mridulm** [comment]
Wont this change not cause issues in history server when handling applications across spark versions ?
`metricToOffset` will end up being different after this change - and depending on the history server version, it will not be able to handle event files from the newer (or older) spark version.

Thoughts @srowen, @dongjoon-hyun, @Hisoka-X  ?

**Comment by mridulm** [comment]
Looking at this more, this should not be that bad an issue as I initially expected - it will lead to older SHS not seeing the newer metric, but not result in failures.
`ExecutorMetrics(Map[String, Long])` constructor and `JSONProtocol.executorMetricsFromJson` , which is used in the deserialization, is resilient to unknown and undefined metrics.

The only issue is, it will map to the wrong index into the array - which can be easily fixed in this PR by not inserting into the middle of the array, but append to end of the `name`.

File: core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala [hunk 2]
```diff
@@ -139,14 +143,17 @@ case object GarbageCollectionMetrics extends ExecutorMetricType with Logging {
   override private[spark] def getMetricValues(memoryManager: MemoryManager): Array[Long] = {
     val gcMetrics = new Array[Long](names.length)
     val mxBeans = ManagementFactory.getGarbageCollectorMXBeans.asScala
-    gcMetrics(4) = mxBeans.map(_.getCollectionTime).sum
+    gcMetrics(6) = mxBeans.map(_.getCollectionTime).sum
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala:146]: With the name ordering change, revert this ?

File: core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala [hunk 3]
```diff
@@ -139,14 +143,17 @@ case object GarbageCollectionMetrics extends ExecutorMetricType with Logging {
...
     mxBeans.foreach { mxBean =>
       if (youngGenerationGarbageCollector.contains(mxBean.getName)) {
         gcMetrics(0) = mxBean.getCollectionCount
         gcMetrics(1) = mxBean.getCollectionTime
       } else if (oldGenerationGarbageCollector.contains(mxBean.getName)) {
         gcMetrics(2) = mxBean.getCollectionCount
         gcMetrics(3) = mxBean.getCollectionTime
+      } else if (BUILTIN_CONCURRENT_GARBAGE_COLLECTOR.equals(mxBean.getName)) {
+        gcMetrics(4) = mxBean.getCollectionCount
+        gcMetrics(5) = mxBean.getCollectionTime
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala:156]: 4 -> 5, 5 -> 6 ?

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Merged to master.
Thanks for fixing this @Hisoka-X !
Thanks for the reviews @srowen, @dongjoon-hyun :-)

**Comment by dongjoon-hyun** [comment]
Thank you, @Hisoka-X , @mridulm and all.

**Pull request closed without merging**